xwiki-platform-core/xwiki-platform-menu/xwiki-platform-menu-ui/src/main/resources/Menu/MenuMacro.xml

```diff
@@ -1159,19 +1159,25 @@ $menuTemplateDoc.content.replace('{', '')
       <code>{{velocity}}
 #set ($id = $xcontext.macro.params.id)
 #set ($type = $xcontext.macro.params.type)
+#set ($label = $xcontext.macro.params.label)
 #set ($colorTheme = $xwiki.getUserPreference('colorTheme'))
 #if ("$!colorTheme" != '')
   ## Make sure we use an absolute reference (see XWIKI-9672)
   #set ($colorTheme = $services.model.resolveDocument($colorTheme, $doc.documentReference))
 #end
 #set ($discard = $xwiki.ssx.use("$xcontext.macro.doc.prefixedFullName", {'colorTheme': $colorTheme}))
 #set ($discard = $xwiki.jsx.use("$xcontext.macro.doc.prefixedFullName"))
+## Make sure the label is non-empty as otherwise the aria-label doesn't work.
+#if ($stringtool.isBlank("$!label"))
+  #set ($label = $wikimacro.context.getXDOM().getIdGenerator().generateUniqueId("MenuLabel"))
```

Comment:
While using the id generator sounds good to get a unique label, I'm wondering if we shouldn't use something more readable/pronounceable here (or maybe just a number?) instead of "IMenuLabel" - "Top navigation menu: IMenuLabel" just doesn't sound good. Also, I'm not sure you can rely on the id generator always being present, so maybe have a fallback for when the call fails.

Also, as there can be just one top navigation menu, wouldn't it be better to just omit the label if there is none provided instead of appending this auto-generated string?

Reply:
> While using the id generator sounds good to get a unique label, I'm wondering if we shouldn't use something more readable/pronounceable here (or maybe just a number?) instead of "IMenuLabel" 

Yesterday with @vmassol  we checked the idGenerator methods, and using 
.generateUniqueId('Menu ', '') 
should give IDs like: Menu 1, Menu 2, Menu 3,...
> Also, I'm not sure you can rely on the id generator always being present, so maybe have a fallback for when the call fails.

The id generator was already used a few lines under there: https://github.com/xwiki/xwiki-platform/blob/cbe530642d81c313d385b3f8401c0056250e1f96/xwiki-platform-core/xwiki-platform-menu/xwiki-platform-menu-ui/src/main/resources/Menu/MenuMacro.xml#L1177 without a fallback so I think it's okay to use it there too.
> Also, as there can be just one top navigation menu, wouldn't it be better to just omit the label if there is none provided instead of appending this auto-generated string?


Navigation blocks don't *need* to have a unique label in all cases:
> If a page includes more than one navigation landmark, each should have a unique label.

from [ARIA Patterns](https://www.w3.org/WAI/ARIA/apg/patterns/landmarks/examples/navigation.html)
However we have another navigation landmark on all pages, which is the [navbar](https://github.com/xwiki/xwiki-platform/blob/649d442f6879127cc57c4169a48aaf51fb2eedcf/xwiki-platform-core/xwiki-platform-flamingo/xwiki-platform-flamingo-skin/xwiki-platform-flamingo-skin-resources/src/main/resources/flamingo/menus_macros.vm#L34). Moreover, at least on view pages there should soon be additional navs in the context of [XWIKI-20696](https://jira.xwiki.org/browse/XWIKI-20696).

Reply:
I think you misunderstood my comment regarding the omission of the label. What I meant is to omit the user-provided label and instead just use, e.g., "Top navigation menu" as label instead of "Top navigation menu: Menu 2". Now if there can indeed be several top navigation menus, this doesn't make that much sense.

> .generateUniqueId('Menu ', '') should give IDs like: Menu 1, Menu 2, Menu 3,...

I think this will rather give an `IllegalArgumentException` as `StringUtils.isAlpha(prefix)` will be `false` as the provided prefix contains a space. You could try `.generateUniqueId('Menu', '')` to get output like `Menu`, `Menu-1`, `Menu-2` etc.

Reply:
>  You could try .generateUniqueId('Menu', '')

Yes, this is what we had discussed actually. Lucas made a mistake in his reply.

Reply:
Sounds ok. 

We might want to improve the id generation implementation to start at "2" when a suffix is added (to get `Menu` and `Menu-2`), and possibly also add a new API signature to force using the suffix (so that we can have `Menu-1` and `Menu-2`). IMO this can be done separately from this PR.

Note that IMO, you also need to improve the Menu application so that it uses the `label` parameter by default / allow the user to set it (it could be an advanced feature without a specific UI for it and that advanced users could do by editing using the object editor for example, but documented).

Reply:
> We might want to improve the id generation implementation to start at "2" when a suffix is added (to get `Menu` and `Menu-2`)

That's not possible/a really breaking change as it will invalidate all links that point to ids (e.g., from headings) with the `-1`-suffix.

Reply:
Yes, it would need to be a different generator or a feature flag for the old generator (a new boolean parameter).